Bluetooth: Classic: Add role change event cb

MarkWangChinese opened the pull request:
add `role_changed` to `struct bt_conn_cb` to notify the HCI_Role_Change event to application.
add BT_HCI_OP_SWITCH_ROLE and struct bt_hci_cp_switch_role.
~add br switch-role shell cmd and test the role_changed callback.~

MarkWangChinese committed:
bluetooth: classic: add role changed callback

add `role_changed` to `struct bt_conn_cb` to notify the HCI_Role_Change
event to application.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

lylezhu2012 commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -478,6 +478,10 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state);
 
 void bt_conn_connected(struct bt_conn *conn);
 
+#if defined(CONFIG_BT_CLASSIC)
```
As my understand, function declarations do not need to be controlled by conditional compilation.

MarkWangChinese replied:
I see similar code as https://github.com/zephyrproject-rtos/zephyr/pull/90126/files#diff-58c6dddbce7d7fd58c9999b995f884c5d46d1f4c79cb17288167595afb389b6aL525

jhedberg replied:
@MarkWangChinese thanks! @lylezhu2012 is correct. If you have the time, please submit a PR to remove those other #ifdefs too. It's a bad idea to have that in header files since it eliminates the possibility of using `IS_ENABLED()` in the calling code.

MarkWangChinese replied:
Thanks for the explain. I will do it.

MarkWangChinese replied:
Fixed this PR.

MarkWangChinese commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -989,6 +990,80 @@ static int cmd_clear(const struct shell *sh, size_t argc, char *argv[])
+	if (err) {
+		shell_error(sh, "Failed to get info (err %d)", err);
+		return err;
+	}
```
Updated.

MarkWangChinese commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -989,6 +990,80 @@ static int cmd_clear(const struct shell *sh, size_t argc, char *argv[])
+	if (buf != NULL) {
+		cp = net_buf_add(buf, sizeof(*cp));
+		memcpy(&cp->bdaddr, &info.br.dst->val[0], sizeof(cp->bdaddr));
+		cp->role = role;
```
From my test, the controller gives error for the result.

MarkWangChinese commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1041,6 +1116,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(br_cmds,
 	SHELL_CMD_ARG(oob, NULL, NULL, cmd_oob, 1, 0),
 	SHELL_CMD_ARG(pscan, NULL, "<value: on, off>", cmd_connectable, 2, 0),
 	SHELL_CMD_ARG(sdp-find, NULL, "<HFPAG, HFPHF>", cmd_sdp_find_record, 2, 0),
+	SHELL_CMD_ARG(switch-role, NULL, "<central, peripheral", cmd_switch_role, 2, 0),
```
Fixed

MarkWangChinese commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -989,6 +990,80 @@ static int cmd_clear(const struct shell *sh, size_t argc, char *argv[])
+
+	if (!strcmp(action, "central")) {
+		/* 0 - Change own Role to Central for this BD_ADDR. */
+		role = 0;
```
Updated.

lylezhu2012 commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1975,6 +1975,18 @@ struct bt_conn_cb {
+	/** @brief The role of the connection has changed.
+	 *
+	 *  This callback notifies the application that the role of the
+	 *  connection has changed.
```
Suggest changing to `This callback notifies the application that the role switch procedure has completed.`

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -989,6 +990,80 @@ static int cmd_clear(const struct shell *sh, size_t argc, char *argv[])
+		}
+	} else {
+		shell_error(sh, "hi cmd fail to create");
+	}
```
Well. I suggest remove this shell command. And adding it when the API is added.

MarkWangChinese replied:
OK, Even I don't agree it, but I still removed the added shell test codes.

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1975,6 +1975,17 @@ struct bt_conn_cb {
+	 *  @param conn Connection object.
+	 *  @param status HCI status of role change event.
+	 */
+	void (*role_changed)(struct bt_conn *conn, uint8_t status);
```
That said, do we even have an API to query the current role of a connection?

MarkWangChinese replied:
`bt_conn_get_info` can get the current role.

jhedberg replied:
Right, maybe that's sufficient then. Another question: why do you bother including a status parameter? If a role change fails then then role didn't change, so why would you call the callback then?

MarkWangChinese replied:
If the role switch is initiated by the DUT, the DUT needs to know the result whether it success or fail. If the role switch is initiated by peer side, this callback can give the indication that the peer tries to switch role even `status` is fail. So this callback is defined as `HCI status of role change event`.

MarkWangChinese replied:
`If the role switch is initiated by peer side, this callback can give the indication that the peer tries to switch role even `status` is fail. ` I am sorry that I think there is no this case. There is no role change event in DUT in this case.

MarkWangChinese replied:
One HCI cmd can be created and sent by `bt_hci_cmd_send_sync`. @lylezhu2012 suggested to add one API and I think it is better. The new API can be added in further pr.

gzh-terry replied:
I would recommend that the HCI API only used for vendor specific HCIs.
Any standard HCI should be handled in stack.

MarkWangChinese replied:
As I said that I fully agree to add one API. But I am afraid that I don't think the callback can only make sense when the new API exist. Because the role switch can be initialized from the peer (it can run any bluetooth stack), then DUT (run zephyr stack) needs the callback to know the role is changed (when role switch success, the DUT host receive the HCI event from controller). Without the callback, the DUT application can only poll the conn info to know the role is changed if DUT application want to know it.

jhedberg replied:
@MarkWangChinese I mean that the callback as it's currently defined in this PR is fine (makes sense) as long as it's combined with an API to initiate the role switch from our side, i.e. I'd propose to add a new commit here to introduce `bt_conn_br_change_role()` or similar.

MarkWangChinese replied:
Hi @jhedberg I have added the new API. I prefer the API as `bt_conn_br_switch_role(const struct bt_conn *conn, uint8_t role)` because the application want to switch as dedicated role not change the current role. For example: application wants to switch as central after connected by peer, give the dedicated role to stack/controller in avoid that the peer initiate the role switch too at the same time, I think the controller can handle the corner case.

MarkWangChinese committed:
bluetooth: classic: add role switch API

add bt_conn_br_switch_role and bt_conn_br_set_role_switchable to control
the role switch, add DEFAULT_ROLE_SWITCHABLE Kconfig to control the default
role switch state.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -558,6 +559,43 @@ struct bt_hci_cp_io_capability_neg_reply {
+#define BT_HCI_OP_WRITE_LINK_POLICY_SETTINGS    BT_OP(BT_OGF_LINK_POLICY, 0x000d)
+struct bt_hci_cp_write_link_policy_settings {
+	uint16_t handle;
+	uint16_t  link_policy_settings;
```
Same here

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -558,6 +559,43 @@ struct bt_hci_cp_io_capability_neg_reply {
+#define BT_HCI_OP_READ_DEFAULT_LINK_POLICY_SETTINGS  BT_OP(BT_OGF_LINK_POLICY, 0x000e)
+struct bt_hci_rp_read_default_link_policy_settings {
+	uint8_t  status;
+	uint16_t  default_link_policy_settings;
```
And here

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -558,6 +559,43 @@ struct bt_hci_cp_io_capability_neg_reply {
+
+#define BT_HCI_OP_WRITE_DEFAULT_LINK_POLICY_SETTINGS BT_OP(BT_OGF_LINK_POLICY, 0x000f)
+struct bt_hci_cp_write_default_link_policy_settings {
+	uint16_t  default_link_policy_settings;
```
And here

jhedberg commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -508,6 +508,11 @@ config BT_COD
+config BT_DEFAULT_ROLE_SWITCHABLE
+	bool "Default Bluetooth Role Switch Enable/Disable State"
+	help
+	  This option sets the controller's default link policy to enable/disable the role switch.
```
Please use similar line width as for help texts for other options, like `BT_COD` above.

jhedberg commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -907,6 +911,42 @@ int bt_br_init(void)
 		}
 	}
 
+	buf = bt_hci_cmd_create(BT_HCI_OP_READ_DEFAULT_LINK_POLICY_SETTINGS, 0);
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_READ_DEFAULT_LINK_POLICY_SETTINGS, buf, &rsp);
+	if (err) {
+		return err;
+	}
```
You don't need `bt_hci_cmd_create()` if you don't have any command parameters. The above is the same as:
```c
err = bt_hci_cmd_send_sync(BT_HCI_OP_READ_DEFAULT_LINK_POLICY_SETTINGS, NULL, &rsp);
```

jhedberg commented on subsys/bluetooth/host/classic/conn_br.c:
```diff
@@ -170,3 +170,107 @@ void bt_br_acl_recv(struct bt_conn *conn, struct net_buf *buf, bool complete)
+
+int bt_conn_br_switch_role(const struct bt_conn *conn, uint8_t role)
+{
+	struct net_buf *buf = NULL;
```
Unnecessary initialization upon declaration.

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -2688,6 +2688,28 @@ struct bt_conn *bt_conn_lookup_addr_br(const bt_addr_t *peer);
+/** @brief Change the role of the conn.
+ *
+ *  @param conn Connection object.
+ *  @param role The role that want to switch as.
```
You should mention what the possible values are. `conn.h` is not an HCI API, so it's not clear if you mean HCI values or something else. And if you mean HCI values you should state where these are defined (`hci_types.h` perhaps?).

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -2688,6 +2688,28 @@ struct bt_conn *bt_conn_lookup_addr_br(const bt_addr_t *peer);
+ *  @return -ENOBUFS HCI command buffer is not available.
+ *  @return -EINVAL @p conn is not a valid @ref BT_CONN_TYPE_BR connection.
+ */
+int bt_conn_br_set_role_switchable(const struct bt_conn *conn, bool enable);
```
Is "switchable" a term that's found in the core spec? If not, I'd avoid introducing it in the API. Perhaps `bt_conn_br_role_switch_allowed()`?

MarkWangChinese replied:
It is not from spec.
Is it better to use `bt_conn_br_set_role_switch_enable(const struct bt_conn *conn, bool enable)` like `bt_conn_le_set_tx_power_report_enable` and `bt_conn_le_set_path_loss_mon_enable`?

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -558,6 +559,43 @@ struct bt_hci_cp_io_capability_neg_reply {
 
+#define BT_HCI_OP_SWITCH_ROLE                   BT_OP(BT_OGF_LINK_POLICY, 0x000b)
+struct bt_hci_cp_switch_role {
+	bt_addr_t bdaddr;
```
Any idea why this is taking bdaddr and not a handle?

MarkWangChinese replied:
The HCI cmd is used for connection, honestly, I don't know why it doesn't use connection handle.

jhedberg commented on subsys/bluetooth/host/classic/conn_br.c:
```diff
@@ -170,3 +170,107 @@ void bt_br_acl_recv(struct bt_conn *conn, struct net_buf *buf, bool complete)
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->handle = conn->handle;
```
Missing `sys_cpu_to_le16()`

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -558,6 +559,43 @@ struct bt_hci_cp_io_capability_neg_reply {
+	uint8_t   role;
+} __packed;
+
+#define BT_HCI_LINK_POLICY_SETTINGS_ENABLE_ROLE_SWITCH  0x01
+#define BT_HCI_LINK_POLICY_SETTINGS_ENABLE_HOLD_MODE    0x02
+#define BT_HCI_LINK_POLICY_SETTINGS_ENABLE_SNIFF_SWITCH 0x04
```
Using `BIT(0)`, `BIT(1)` and `BIT(2)` would be clearer/

jhedberg commented on subsys/bluetooth/host/classic/conn_br.c:
```diff
@@ -170,3 +170,107 @@ void bt_br_acl_recv(struct bt_conn *conn, struct net_buf *buf, bool complete)
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->handle = conn->handle;
```
Missing `sys_cpu_to_le16()`

jhedberg commented on subsys/bluetooth/host/classic/conn_br.c:
```diff
@@ -170,3 +170,107 @@ void bt_br_acl_recv(struct bt_conn *conn, struct net_buf *buf, bool complete)
+		return err;
+	}
+
+	is_enabled = link_policy_settings & BT_HCI_LINK_POLICY_SETTINGS_ENABLE_ROLE_SWITCH;
+
+	if (enable == is_enabled) {
```
To make it clear that you mean `&` and not `&&` I think it's better to put the operation inside parenthesis. Also, no empty line between the assignment and the test, please (this is also inconsistent with yourself, since you don't do it e.g. on line 261/262 either).

jhedberg commented on subsys/bluetooth/host/classic/conn_br.c:
```diff
@@ -170,3 +170,107 @@ void bt_br_acl_recv(struct bt_conn *conn, struct net_buf *buf, bool complete)
+	}
+
+	link_policy_settings ^= BT_HCI_LINK_POLICY_SETTINGS_ENABLE_ROLE_SWITCH;
+	err = bt_conn_br_write_link_policy_settings(conn, link_policy_settings);
+
+	return err;
```
I'd just replace this with `return bt_conn_br_write_...();`

jhedberg commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -907,6 +911,42 @@ int bt_br_init(void)
+	default_link_policy_settings = rp->default_link_policy_settings;
+
+	bool should_enable = CONFIG_BT_DEFAULT_ROLE_SWITCHABLE;
+	bool is_enabled = default_link_policy_settings &
+			  BT_HCI_LINK_POLICY_SETTINGS_ENABLE_ROLE_SWITCH;
```
Use `()` here. Added benefit is that editors will do the split line alignment more correctly.

jhedberg commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -801,7 +801,11 @@ int bt_br_init(void)
 	struct bt_hci_write_local_name *name_cp;
 	struct bt_hci_cp_write_class_of_device *cod;
+	struct bt_hci_rp_read_default_link_policy_settings *rp;
+	struct bt_hci_cp_write_default_link_policy_settings *policy_cp;
```
Move this into the if-branch where it's actually used (variables should always be declared in the smallest possible scope)

jhedberg reviewed (changes requested):
A couple of initial comments. I'll do a more thorough review in the next iteration.

lylezhu2012 commented on subsys/bluetooth/host/classic/conn_br.c:
```diff
@@ -170,3 +170,104 @@ void bt_br_acl_recv(struct bt_conn *conn, struct net_buf *buf, bool complete)
+	return bt_hci_cmd_send_sync(BT_HCI_OP_WRITE_LINK_POLICY_SETTINGS, buf, NULL);
+}
+
+int bt_conn_br_set_role_switch_enable(const struct bt_conn *conn, bool enable)
```
Whether is it better to change function name to `bt_conn_br_role_switch`?
```suggestion
int bt_conn_br_role_switch(const struct bt_conn *conn, bool enable)
```

MarkWangChinese replied:
It is talked here https://github.com/zephyrproject-rtos/zephyr/pull/90126#discussion_r2151499920 and it is more aligned with the other bluetooth APIs.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1207,6 +1207,86 @@ static int cmd_info(const struct shell *sh, size_t argc, char *argv[])
+
+	action = argv[1];
+
+	if (!strcmp(action, "enable")) {
+		enable = true;
+	} else if (!strcmp(action, "disable")) {
+		enable = false;
+	} else {
+		shell_help(sh);
+		return SHELL_CMD_HELP_PRINTED;
+	}
```
There is a function `shell_strtobool()`. It can be used to parse the argument from string to bool.

MarkWangChinese commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -681,22 +681,22 @@ void bt_hci_role_change(struct net_buf *buf)
-	} else {
-		conn->role = BT_CONN_ROLE_CENTRAL;
+	if (!evt->status) {
+		if (evt->role) {
```
Updated with the HCI role MACRO.

lylezhu2012 commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -5180,6 +5186,9 @@ SHELL_STATIC_SUBCMD_SET_CREATE(bt_cmds,
 #endif
 #endif /* CONFIG_BT_SMP || CONFIG_BT_CLASSIC) */
 #endif /* CONFIG_BT_CONN */
+#if defined(CONFIG_BT_HCI_MESH_EXT)
```
Why is it changed here?
